src/net_processing.cpp

```diff
@@ -2566,6 +2572,33 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
         return;
     }
 
+    if (msg_type == NetMsgType::BLOCKRELAY) {
+        if (pfrom.fSuccessfullyConnected) {
+            // Disconnect peers that send blockrelay message after VERACK; this
+            // must be negotiated between VERSION and VERACK.
+            pfrom.fDisconnect = true;
+            return;
+        }
+        if (pfrom.IsFullOutboundConn() || pfrom.IsAddrFetchConn()) {
+            // If we picked an outbound for tx- or addr-relay and it sends us a
+            // BLOCKRELAY, find another peer.
+            pfrom.fDisconnect = true;
+            return;
+        }
+        if (pfrom.m_tx_relay != nullptr && WITH_LOCK(pfrom.m_tx_relay->cs_filter, return pfrom.m_tx_relay->fRelayTxes)) {
+            // Can't send a BLOCKRELAY message if they didn't turn off tx-relay
+            // in the version message.
+            pfrom.fDisconnect = true;
+        }
+        if (pfrom.IsInboundConn()) {
+            pfrom.UpdateConnectionType(ConnectionType::INBOUND_BLOCK_RELAY);
+        }
+        // If an outbound peer (manual or block-relay-only) sends us this
+        // message, we just ignore it. We won't relay transactions with such
```

Comment:
We do relay transactions with manual peers, right?

Reply:
I should update the comment to be clearer. What I meant to describe was that we only disconnect an outbound peer who sends us the BLOCKRELAY message if we expect them to be a tx-relay or addr-relay peer.  If we are connecting to them as a MANUAL connection, then we'll stay connected and just not relay transactions on that link.

In addition to fixing the comment, I could improve the code to ensure we don't advertise addrs to such peers either -- I think that will be easy to do at the same time that I ensure that a BLOCKRELAY peer doesn't send us other disallowed messages (mempool, filterload/add/clear, etc).

Reply:
Would some users be using `-addnode` to start with a trusted peer that can then also provide addrs? If so, when we receive `BLOCKRELAY` from a `MANUAL` connection, and if addrman is empty, is it worth alerting the user by logging?